neurow/lib/neurow/public_api/endpoint.ex

```diff
@@ -194,58 +196,84 @@ defmodule Neurow.PublicApi.Endpoint do
-        conn |> loop(sse_timeout, keep_alive, new_last_message, new_last_message, jwt_exp)
+  def loop(conn, sse_timeout_ms, keep_alive_ms, last_message_ts, last_ping_ts, jwt_exp) do
+    now_ms = :os.system_time(:millisecond)
 
-      # Consume useless messages to avoid memory overflow
-      _ ->
-        conn |> loop(sse_timeout, keep_alive, last_message, last_ping, jwt_exp)
-    after
-      1000 ->
-        now_ms = :os.system_time(:millisecond)
-
-        cond do
-          # SSE Timeout
-          now_ms - last_message > sse_timeout ->
-            Logger.debug("Client disconnected due to inactivity")
-            conn |> write_raw_chunk("event: timeout")
-
-          # SSE Keep alive, send a ping
-          now_ms - last_ping > keep_alive ->
-            conn
-            |> write_raw_chunk("event: ping")
-            |> loop(sse_timeout, keep_alive, last_message, now_ms, jwt_exp)
+    cond do
+      # SSE Timeout
+      now_ms - last_message_ts > sse_timeout_ms ->
+        Logger.info("Client disconnected due to inactivity")
+        conn |> write_chunk("event: timeout")
+
+      # SSE Keep alive, send a ping
+      now_ms - last_ping_ts > keep_alive_ms ->
+        conn
+        |> write_chunk("event: ping")
+        |> loop(
+          sse_timeout_ms,
+          keep_alive_ms,
+          last_message_ts,
+          now_ms,
+          jwt_exp
+        )
+
+      # JWT token expired
+      jwt_exp * 1000 < now_ms ->
+        conn |> write_chunk("event: credentials_expired")
+
+      # Otherwise, let's wait for a message or the next tick
+      true ->
+        # Compute the waiting time before the next tick
+        next_ping_ms = last_ping_ts + keep_alive_ms - now_ms
+        timeout_ms = last_message_ts + sse_timeout_ms - now_ms
+        jwt_exp_ms = jwt_exp * 1000 - now_ms
+
+        # The Erlang process scheduler does not guarantee the `after` block will be executed with a ms precision,
+        # So a small tolerance is added, also a minimum of 100ms is set to avoid busy waiting
+        next_tick_ms = max(Enum.min([next_ping_ms, timeout_ms, jwt_exp_ms]) + 20, 100)
+
+        receive do
+          {:pubsub_message, message} ->
+            conn = write_chunk(conn, message)
+            Stats.inc_msg_published()
+            new_last_message_ts = :os.system_time(:millisecond)
 
-          # JWT token expired
-          jwt_exp * 1000 < now_ms ->
-            conn |> write_raw_chunk("event: credentials_expired")
+            conn
+            |> loop(
+              sse_timeout_ms,
+              keep_alive_ms,
+              new_last_message_ts,
+              new_last_message_ts,
+              jwt_exp
+            )
 
-          # We need to stop
-          StopListener.close_connections?() ->
-            conn |> write_raw_chunk("event: reconnect")
+          {:shutdown, _value} ->
+            Logger.debug("Client on disconnected due to node shutdown")
+            conn |> write_chunk("event: node_shutdown")
```

Comment:
Can we use a terminology for client?

Reply:
What do you mean ? You prefer "reconnect" ?